clippy_lints/src/include_file_outside_project.rs

```diff
@@ -0,0 +1,199 @@
+use rustc_ast::{Attribute, LitKind, MetaItem, MetaItemInner};
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::{Expr, ExprKind, HirId, Item};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::{FileName, Span, sym};
+
+use clippy_config::Conf;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::macros::root_macro_call_first_node;
+
+use cargo_metadata::MetadataCommand;
+
+use std::path::{Path, PathBuf};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check if files included with one of the `include` macros (ie, `include!`, `include_bytes!`
+    /// and `include_str!`) or the `path` attribute are actually part of the project.
+    ///
+    /// ### Why is this bad?
+    /// If the included file is outside of the project folder, it will not be part of the releases,
+    /// prevent project to work when others use it.
+    ///
+    /// ### Example
+    /// ```ignore
+    /// let x = include_str!("/etc/passwd");
+    ///
+    /// #[path = "/etc/passwd"]
+    /// mod bar;
+    /// ```
+    /// Use instead:
+    /// ```ignore
+    /// let x = include_str!("./passwd");
+    ///
+    /// #[path = "./passwd"]
+    /// mod bar;
+    /// ```
+    #[clippy::version = "1.84.0"]
+    pub INCLUDE_FILE_OUTSIDE_PROJECT,
+    suspicious,
+    "checks that all included files are inside the project folder"
+}
+
+pub(crate) struct IncludeFileOutsideProject {
+    cargo_manifest_dir: Option<PathBuf>,
+    warned_spans: FxHashSet<PathBuf>,
+    can_check_crate: bool,
+}
+
+impl_lint_pass!(IncludeFileOutsideProject => [INCLUDE_FILE_OUTSIDE_PROJECT]);
+
+impl IncludeFileOutsideProject {
+    pub(crate) fn new(conf: &'static Conf) -> Self {
+        let mut can_check_crate = true;
+        if !conf.cargo_ignore_publish {
+            match MetadataCommand::new().no_deps().exec() {
+                Ok(metadata) => {
+                    for package in &metadata.packages {
+                        // only run the lint if publish is `None` (`publish = true` or skipped entirely)
+                        // or if the vector isn't empty (`publish = ["something"]`)
+                        if !matches!(package.publish.as_deref(), Some([]) | None) {
+                            can_check_crate = false;
+                            break;
+                        }
+                    }
```

Comment:
This would have to be a `cargo` lint if it's doing this to deduplicate the `cargo metadata` invocation and so it doesn't happen by default

But I don't think we need to here, `cargo publish` already copies the project files into `target/package/crate-x.y.z` before checking it builds, which would catch most `../../outside-the-project` includes

The case that it wouldn't catch is an include of an absolute path, but it seems reasonable to lint these unconditionally